File: leetcode/0019-remove-nth-node-from-end-of-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        lead = head
        while n > 0:
            lead = lead.next
            n = n - 1
        
        lag = head
        prev = None
        while lead != None:
            lead = lead.next
            prev = lag
            lag = lag.next
        
        if prev == None:
            return head.next
        
        prev.next = lag.next
        lag.next = None
        
        return head

File: leetcode/0019-remove-nth-node-from-end-of-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* aheadPtr = head;
        n--;
        while (n > 0) {
            aheadPtr = aheadPtr->next;
            n--;
        }
        
        if (aheadPtr->next == nullptr) {
            ListNode* newHead = head->next;
            delete head;
            return newHead;
        }
        
        ListNode* behindPtr = head;
        ListNode* prevPtr = nullptr;
        while (aheadPtr->next != nullptr) {
            prevPtr = behindPtr;
            aheadPtr = aheadPtr->next;
            behindPtr = behindPtr->next;
        }
        
        prevPtr->next = behindPtr->next;
        
        delete behindPtr;
        
        return head;
    }
};